### PR TITLE
feat(argo-cd): Support livenessProbe for Application Controller

### PR DESCRIPTION
**Pull request by yu-croco**
Resolves https://github.com/argoproj/argo-helm/issues/2973

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

### CODE REVIEW

File: charts/argo-cd/templates/argocd-application-controller/deployment.yaml [hunk 1]
```diff
@@ -310,6 +310,17 @@ spec:
           timeoutSeconds: {{ .Values.controller.readinessProbe.timeoutSeconds }}
           successThreshold: {{ .Values.controller.readinessProbe.successThreshold }}
           failureThreshold: {{ .Values.controller.readinessProbe.failureThreshold }}
+        {{- if .Values.controller.livenessProbe.enabled }}
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: metrics
```
Comment (yu-croco) [inline charts/argo-cd/templates/argocd-application-controller/deployment.yaml:317]: same as `readinessProbe` 📝 
Ref: https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/templates/argocd-application-controller/deployment.yaml#L304-L307

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml [hunk 2]
```diff
@@ -309,6 +309,17 @@ spec:
           timeoutSeconds: {{ .Values.controller.readinessProbe.timeoutSeconds }}
           successThreshold: {{ .Values.controller.readinessProbe.successThreshold }}
           failureThreshold: {{ .Values.controller.readinessProbe.failureThreshold }}
+        {{- if .Values.controller.livenessProbe.enabled }}
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: metrics
```
Comment (yu-croco) [inline charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:316]: same as readinessProbe 📝
Ref: https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/templates/argocd-application-controller/statefulset.yaml#L303-L306

**Review by mbevc1 — APPROVED** [review]

**Review by mkilchhofer — COMMENTED** [review]
We should place a (prominent) note that this can have side effects.

We once had the probes on the application-controller but due to an (upstream) issue, upstream removed it and thus we also removed it in the helm chart.

> Current controller liveness probe is [a noop](https://github.com/argoproj/argo-cd/blob/master/controller/appcontroller.go#L197) just checking if the metric-server is able to receive requests.
> 
> In cases when the controller is overloaded reconciling large queues restarting the Pod is more harmful than letting it running. In discussion with @alexmt we understand that liveness probe should be removed from the controller to prevent it from being restarted.
> 
> Signed-off-by: Leonardo Luz Almeida [leonardo_almeida@intuit.com](mailto:leonardo_almeida@intuit.com)

Xref:
- https://github.com/argoproj/argo-cd/pull/9557
- https://github.com/argoproj/argo-helm/pull/1581

**Comment by yu-croco** [comment]
@mkilchhofer Thank you for your review with the backgroud.
It seems that it's not worth to take risk even it's optional... 🤔

**Pull request closed without merging**